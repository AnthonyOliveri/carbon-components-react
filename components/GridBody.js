import GridBodyRow from './GridBodyRow';
import React from 'react';
import PropTypes from 'prop-types';

export default function GridBody(props) {
  const { columns, expandedRowIndices } = props;
  const dataRows = props.data.map((item, index) => {
    const rowExpanded =
      expandedRowIndices && expandedRowIndices.indexOf(index) > -1;

    return (
      <GridBodyRow
        className={rowExpanded ? 'bx--grid-body-row--expanded' : ''}
        columns={columns}
        data={item}
        index={index}
        key={index}
        renderRowExpansion={props.renderRowExpansion}
        rowExpanded={rowExpanded}
      />
    );
  });

  let detailPanel = '';

  if (
    typeof props.expandedDetailIndex === 'number' &&
    props.renderDetailExpansion
  ) {
    detailPanel = (
      <div
        className="bx--grid-body__details"
        style={{ width: `${props.detailPanelWidth}%` }}
      >
        {props.renderDetailExpansion(props.data[props.expandedDetailIndex])}
      </div>
    );
  }

  return (
    <div className="bx--grid-body">
      {dataRows}
      {detailPanel}
    </div>
  );
}

GridBody.defaultProps = {
  columns: [],
  data: [],
  detailPanelWidth: 70,
  expandedDetailIndex: null,
  expandedRowIndices: [],
  renderDetailExpansion: () => {},
  renderRowExpansion: () => {},
};

GridBody.propTypes = {
  columns: PropTypes.array.isRequired,
  data: PropTypes.array.isRequired,
  detailPanelWidth: PropTypes.number,
  expandedDetailIndex: PropTypes.number,
  expandedRowIndices: PropTypes.array,
  renderDetailExpansion: PropTypes.func,
  renderRowExpansion: PropTypes.func,
};
