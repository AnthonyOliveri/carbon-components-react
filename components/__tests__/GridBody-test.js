import GridBody from '../GridBody';
import React from 'react';
import { shallow } from 'enzyme';

jest.unmock('../GridBody');

const shallowRender = props => shallow(<GridBody {...props} />);

describe('GridBody', () => {
  const columns = [
    {
      name: 'column1',
      title: 'column1',
      render: data => data.column1,
      width: 50,
    },
    {
      name: 'column2',
      title: 'column2',
      render: data => data.column2,
      width: 25,
    },
    {
      name: 'column3',
      title: 'column3',
      render: data => data.column3,
      width: 25,
    },
  ];
  describe('check if props are rendered correctly', () => {
    describe('standard body - no row expansion', () => {
      const data = [
        {
          column1: 'stuff1',
          column2: 'more1',
          column3: 'test1',
        },
        {
          column1: 'stuff2',
          column2: 'more2',
          column3: 'test2',
        },
        {
          column1: 'stuff3',
          column2: 'more3',
          column3: 'test3',
        },
      ];
      const wrapper = shallowRender({
        columns,
        data,
        expandedDetailIndex: null,
        expandedRowIndices: [],
        renderDetailExpansion: null,
        renderRowExpansion: null,
      });

      it('renders correctly', () => {
        expect(wrapper).toMatchSnapshot();
      });
    });

    describe('standard body - row 2 expanded', () => {
      const data = [
        {
          column1: 'stuff1',
          column2: 'more1',
          column3: 'test1',
        },
        {
          column1: 'stuff2',
          column2: 'more2',
          column3: 'test2',
          expander: 'expand',
        },
        {
          column1: 'stuff3',
          column2: 'more3',
          column3: 'test3',
        },
      ];
      const wrapper = shallowRender({
        columns,
        data,
        expandedDetailIndex: null,
        expandedRowIndices: [2],
        renderDetailExpansion: (rowData, index) =>
          `${rowData.expander}-${index}`,
        renderRowExpansion: rowData =>
          `${rowData.column1}-${rowData.column3}-${rowData.column2}-${rowData.expander}`,
      });

      it('renders correctly', () => {
        expect(wrapper).toMatchSnapshot();
      });
    });

    describe('standard body - row 2 detail expanded', () => {
      const data = [
        {
          column1: 'stuff1',
          column2: 'more1',
          column3: 'test1',
        },
        {
          column1: 'stuff2',
          column2: 'more2',
          column3: 'test2',
          expander: 'expand',
        },
        {
          column1: 'stuff3',
          column2: 'more3',
          column3: 'test3',
        },
      ];
      const wrapper = shallowRender({
        columns,
        data,
        expandedDetailIndex: 2,
        expandedRowIndices: [],
        renderDetailExpansion: (rowData, index) =>
          `${rowData.expander}-${index}`,
        renderRowExpansion: rowData =>
          `${rowData.column1}-${rowData.column3}-${rowData.column2}-${rowData.expander}`,
      });

      it('renders correctly', () => {
        expect(wrapper).toMatchSnapshot();
      });
    });
  });
});
