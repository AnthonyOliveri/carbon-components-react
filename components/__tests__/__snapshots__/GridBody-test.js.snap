exports[`GridBody check if props are rendered correctly standard body - no row expansion renders correctly 1`] = `
<div
  className="bx--grid-body">
  <GridBodyRow
    className=""
    columns={
      Array [
        Object {
          "name": "column1",
          "render": [Function],
          "title": "column1",
          "width": 50,
        },
        Object {
          "name": "column2",
          "render": [Function],
          "title": "column2",
          "width": 25,
        },
        Object {
          "name": "column3",
          "render": [Function],
          "title": "column3",
          "width": 25,
        },
      ]
    }
    data={
      Object {
        "column1": "stuff1",
        "column2": "more1",
        "column3": "test1",
      }
    }
    index={0}
    renderRowExpansion={null}
    rowExpanded={false} />
  <GridBodyRow
    className=""
    columns={
      Array [
        Object {
          "name": "column1",
          "render": [Function],
          "title": "column1",
          "width": 50,
        },
        Object {
          "name": "column2",
          "render": [Function],
          "title": "column2",
          "width": 25,
        },
        Object {
          "name": "column3",
          "render": [Function],
          "title": "column3",
          "width": 25,
        },
      ]
    }
    data={
      Object {
        "column1": "stuff2",
        "column2": "more2",
        "column3": "test2",
      }
    }
    index={1}
    renderRowExpansion={null}
    rowExpanded={false} />
  <GridBodyRow
    className=""
    columns={
      Array [
        Object {
          "name": "column1",
          "render": [Function],
          "title": "column1",
          "width": 50,
        },
        Object {
          "name": "column2",
          "render": [Function],
          "title": "column2",
          "width": 25,
        },
        Object {
          "name": "column3",
          "render": [Function],
          "title": "column3",
          "width": 25,
        },
      ]
    }
    data={
      Object {
        "column1": "stuff3",
        "column2": "more3",
        "column3": "test3",
      }
    }
    index={2}
    renderRowExpansion={null}
    rowExpanded={false} />
</div>
`;

exports[`GridBody check if props are rendered correctly standard body - row 2 detail expanded renders correctly 1`] = `
<div
  className="bx--grid-body">
  <GridBodyRow
    className=""
    columns={
      Array [
        Object {
          "name": "column1",
          "render": [Function],
          "title": "column1",
          "width": 50,
        },
        Object {
          "name": "column2",
          "render": [Function],
          "title": "column2",
          "width": 25,
        },
        Object {
          "name": "column3",
          "render": [Function],
          "title": "column3",
          "width": 25,
        },
      ]
    }
    data={
      Object {
        "column1": "stuff1",
        "column2": "more1",
        "column3": "test1",
      }
    }
    index={0}
    renderRowExpansion={[Function]}
    rowExpanded={false} />
  <GridBodyRow
    className=""
    columns={
      Array [
        Object {
          "name": "column1",
          "render": [Function],
          "title": "column1",
          "width": 50,
        },
        Object {
          "name": "column2",
          "render": [Function],
          "title": "column2",
          "width": 25,
        },
        Object {
          "name": "column3",
          "render": [Function],
          "title": "column3",
          "width": 25,
        },
      ]
    }
    data={
      Object {
        "column1": "stuff2",
        "column2": "more2",
        "column3": "test2",
        "expander": "expand",
      }
    }
    index={1}
    renderRowExpansion={[Function]}
    rowExpanded={false} />
  <GridBodyRow
    className=""
    columns={
      Array [
        Object {
          "name": "column1",
          "render": [Function],
          "title": "column1",
          "width": 50,
        },
        Object {
          "name": "column2",
          "render": [Function],
          "title": "column2",
          "width": 25,
        },
        Object {
          "name": "column3",
          "render": [Function],
          "title": "column3",
          "width": 25,
        },
      ]
    }
    data={
      Object {
        "column1": "stuff3",
        "column2": "more3",
        "column3": "test3",
      }
    }
    index={2}
    renderRowExpansion={[Function]}
    rowExpanded={false} />
  <div
    className="bx--grid-body__details"
    style={
      Object {
        "width": "70%",
      }
    }>
    undefined-undefined
  </div>
</div>
`;

exports[`GridBody check if props are rendered correctly standard body - row 2 expanded renders correctly 1`] = `
<div
  className="bx--grid-body">
  <GridBodyRow
    className=""
    columns={
      Array [
        Object {
          "name": "column1",
          "render": [Function],
          "title": "column1",
          "width": 50,
        },
        Object {
          "name": "column2",
          "render": [Function],
          "title": "column2",
          "width": 25,
        },
        Object {
          "name": "column3",
          "render": [Function],
          "title": "column3",
          "width": 25,
        },
      ]
    }
    data={
      Object {
        "column1": "stuff1",
        "column2": "more1",
        "column3": "test1",
      }
    }
    index={0}
    renderRowExpansion={[Function]}
    rowExpanded={false} />
  <GridBodyRow
    className=""
    columns={
      Array [
        Object {
          "name": "column1",
          "render": [Function],
          "title": "column1",
          "width": 50,
        },
        Object {
          "name": "column2",
          "render": [Function],
          "title": "column2",
          "width": 25,
        },
        Object {
          "name": "column3",
          "render": [Function],
          "title": "column3",
          "width": 25,
        },
      ]
    }
    data={
      Object {
        "column1": "stuff2",
        "column2": "more2",
        "column3": "test2",
        "expander": "expand",
      }
    }
    index={1}
    renderRowExpansion={[Function]}
    rowExpanded={false} />
  <GridBodyRow
    className="bx--grid-body-row--expanded"
    columns={
      Array [
        Object {
          "name": "column1",
          "render": [Function],
          "title": "column1",
          "width": 50,
        },
        Object {
          "name": "column2",
          "render": [Function],
          "title": "column2",
          "width": 25,
        },
        Object {
          "name": "column3",
          "render": [Function],
          "title": "column3",
          "width": 25,
        },
      ]
    }
    data={
      Object {
        "column1": "stuff3",
        "column2": "more3",
        "column3": "test3",
      }
    }
    index={2}
    renderRowExpansion={[Function]}
    rowExpanded={true} />
</div>
`;
